substreams/ethereum-sdai/src/modules.rs

```diff
@@ -0,0 +1,286 @@
+        .tx_components
+        .iter()
+        .for_each(|tx_component| {
+            let tx = tx_component.tx.as_ref().unwrap();
+            transaction_contract
+                .entry(tx.index)
+                .or_insert_with(|| TransactionChanges::new(tx))
+                .component_changes
+                .extend_from_slice(&tx_component.components);
+        });
+
+    // Balance changes are gathered by the `StoreDelta` based on `PoolBalanceChanged` creating
+    //  `BlockBalanceDeltas`. We essentially just process the changes that occurred to the `store`
+    // this  block. Then, these balance changes are merged onto the existing map of tx contract
+    // changes,  inserting a new one if it doesn't exist.
+    aggregate_balances_changes(balance_store, deltas)
+        .into_iter()
+        .for_each(|(_, (tx, balances))| {
+            transaction_contract
+                .entry(tx.index)
+                .or_insert_with(|| TransactionChanges::new(&tx))
+                .balance_changes
+                .extend(
+                    balances
+                        .into_values()
+                        .flat_map(|inner_map| inner_map.into_values()),
+                );
+        });
+
+    // Extract and insert any storage changes that happened for any of the components.
+    extract_contract_changes(
+        &block,
+        |addr| {
+            components_store
+                .get_last(format!("pool:0x{0}", hex::encode(addr)))
+                .is_some()
+        },
+        &mut transaction_contract,
+    );
+
+    // Process all `transaction_contract_changes` for final output in the `BlockContractChanges`,
+    //  sorted by transaction index (the key).
+    Ok(BlockChanges {
+        block: Some((&block).into()),
+        changes: transaction_contract
+            .drain()
+            .sorted_unstable_by_key(|(index, _)| *index)
+            .filter_map(|(_, change)| {
+                if change.contract_changes.is_empty() &&
+                    change.component_changes.is_empty() &&
+                    change.balance_changes.is_empty()
+                {
+                    None
+                } else {
+                    Some(change)
+                }
+            })
+            .collect::<Vec<_>>(),
+    })
+}
+
+fn is_deployment_tx(tx: &eth::v2::TransactionTrace, vault_address: &[u8]) -> bool {
+    let created_accounts = tx
+        .calls
+        .iter()
+        .flat_map(|call| {
+            call.account_creations
+                .iter()
+                .map(|ac| ac.account.to_owned())
+        })
+        .collect::<Vec<_>>();
+
+    if let Some(deployed_address) = created_accounts.first() {
+        return deployed_address.as_slice() == vault_address;
+    }
+    false
+}
+
+fn find_deployed_underlying_address(vault_address: &[u8]) -> Option<[u8; 20]> {
+    // Ethereum
+    match vault_address {
+        // sDai
+        hex!("83F20F44975D03b1b09e64809B757c47f942BEeA") => {
```

Comment:
Hardcoded address - it won't work on other chains. I think you can get this value from the constructor?